lib/fluent/plugin/output.rb

```diff
@@ -825,6 +822,16 @@ def metadata(tag, time, record)
         end
       end
 
+      def calculate_timekey(time)
+        time_int = time.to_i
+        if @buffer_config.timekey_use_utc
+          (time_int - (time_int % @buffer_config.timekey)).to_i
+        else
+          offset = Time.at(time_int).utc_offset
```

Comment:
Should consider `timekey_zone` value instead of `utc_offset`?

Reply:
Maybe yes. How about the following function?

```diff
diff --git a/lib/fluent/timezone.rb b/lib/fluent/timezone.rb
index bb6fa868..0e79dd03 100644
--- a/lib/fluent/timezone.rb
+++ b/lib/fluent/timezone.rb
@@ -139,5 +139,17 @@ module Fluent
 
       return nil
     end
+
+    def self.utc_offset(time, timezone)
+      return 0 if timezone.nil?
+
+      case timezone
+      when NUMERIC_PATTERN
+        Time.zone_offset(timezone)
+      when NAME_PATTERN
+        tz = TZInfo::Timezone.get(timezone)
+        tz.period_for_utc(time).utc_total_offset
+      end
+    end
   end
 end
```

Reply:
looks good